ibroadcast-uploader.py

```diff
@@ -287,35 +277,94 @@ def calcmd5(self, filePath="."):
         self.__load_md5_int()
         self.__load_md5_ext()
-        skipped = 0
-        failed = 0
-        uploaded = 0
 
-        for filename in self.progressbar(self.files, "Processing: ", 60):
-            if self.be_verbose:
-                print(filename)
-            
+        calculated = False
+        current_path=''
+        file_list = self.files[:]
+        if not self.be_silent and not self.be_verbose:
+            file_list = self.progressbar(self.files[:], "Calculating MD5 hashes: ", 60)
+
+        for filename in file_list:
             # Get an md5 of the file contents and compare it to whats up
             # there already
             if (not self.no_cache) and filename in self.md5_int:
                 file_md5 = self.md5_int[filename]
             else:
-                if self.be_verbose:
-                    print('Calculating MD5 for file...')
+                calculated = True
+                if not self.be_silent and self.be_verbose:
+                    if os.path.dirname(filename) != current_path:
+                        current_path = os.path.dirname(filename)
+                        print('\nChecking directory %s...' % current_path)
+                    print('Calculating MD5 for file %s... ' % os.path.basename(filename), end='')
                 file_md5 = self.calcmd5(filename)
                 self.md5_int[filename] = file_md5
 
             if file_md5 in self.md5_ext:
-                if self.be_verbose:
-                    print('Skipping - already uploaded.')
-                skipped += 1
-                continue
+                self.skipped_files.append(filename)
+                if not self.be_silent and self.be_verbose and calculated:
+                    print('Skipping, already uploaded.')
+                # Removing the files is faster than later comparing
+                # which files are present in self.skipped_files
+                self.files.remove(filename)
+            elif not self.be_silent and self.be_verbose and calculated:
+                print()
+            calculated = False
+
+        with open(self.md5_int_path, 'w') as fp:
+            json.dump(self.md5_int, fp, indent = 2)
+
+    def progressbar(self, it, prefix="", size=60, out=sys.stdout):
+        count = len(it)
+        def show(j):
+            x = int(size*j/count)
+            print("{}[{}{}] {}/{}".format(prefix, "#"*x, "."*(size-x), j, count),
+                end='\r', file=out, flush=True)
+        if (not self.be_verbose) and (not self.be_silent):
+            show(0)
+        for i, item in enumerate(it):
+            yield item
+            if (not self.be_verbose) and (not self.be_silent):
+                show(i+1)
+        if (not self.be_verbose) and (not self.be_silent):
+            print("\n", flush=True, file=out)
+
+    def split_file_list(self, slices):
+        self.files = [self.files[i * slices:(i + 1) * slices] for i in range((len(self.files) + slices - 1) // slices )]
```

Comment:
What this does and why it's being called just once with `slices` = 1? I think this piece of code needs a comment.

Reply:
What this does is split the list into chunks of size `slices`, so when we pass the list to the `upload()` function, only a single file is taken per thread (which is why we call it with the value of 1). I had tried avoiding doing this, but encountered the problem that the whole string was being printed out on screen, letter by letter. And only now I realize what the problem was: I was later iterating over the string inside the `upload()` function; I just had to remove the for loop and reindent the remaining code.